### ISSUE TITLE
[FR] Hyperion.NG on Docker Hub

### ISSUE THREAD

**Issue by Hukuma1** [hyperion-project/hyperion.ng#764]
## Feature request

I know there's containers for actually compiling/cross-compiling Hyperion.NG (https://hub.docker.com/u/hyperionproject), but can this official repository also please provide the current alpha releases for multiple platforms as well? There's a few (outdated) non-official Docker containers with Hyperion.NG out there, but not for all platforms. Currently there's no Raspberry Pi arm version for instance.

#### What problem does this feature solve?
Super easy installation for anyone running VM/RPi Hyperion setups.

#### What does the proposed API look like?
N/A

#### How should this be implemented in your opinion?
Official HyperionNG repo to add containers for all platforms of current pre-releases?

#### Are you willing to work on this yourself?
Yes, but I'm not a coder. Am able to happily test any RPi arm7 containers.

**Comment by Hukuma1** [comment]
Hi. First off, big fan of the project so thank you and keep up the great work!

To clarify, yes, I am not talking about the actual releases that are now being compiled. This is a step further. For people using Docker Containers, it would essentially be a container with that release already installed.

An example of what I'm talking about is here: https://hub.docker.com/r/sirfragalot/hyperion.ng

However the issues with the above repository are: 1) it's outdated, so it's using a compiled version from 3 months ago and 2) will only work on OS/ARCH linux/amd64. So people with Raspberry Pi's will not be able to use this Docker Container for example.

**Comment by Paulchen-Panther** [comment]
Now I understand your question. :-)
Hyperion analyzes the screen content, which is not possible or only possible to a limited extent in a Docker container. This is the reason why we do not deliver finished Docker containers with Hyperion.

That our Docker containers (for the production of our releases) are not up to date, is because we want to be backwards compatible. The releases have the necessary dependencies, which enables execution on upward systems.

**Comment by Hukuma1** [comment]
Maybe I'm confusing myself and you at the same time. Does that mean the above docker hub by SirFragalot would not have a working hyperion executable for AMD64 systems? :(

Sorry just got exposed to Docker Containers and have everything running in one now. Home Assistant, Pi-Hole, etc. Wanted to add Hyperion NG to the stack. Is this something that may be possible later or a limitation of containers and screen capture? Could it work with elevated privileges or anything?

**Comment by Paulchen-Panther** [comment]
> Does that mean the above docker hub by SirFragalot would not have a working hyperion executable for AMD64 systems? :(

It's already running. I only doubt that this Docker container can pick up the screen signal and forward it to Hyperion.



> Is this something that may be possible later or a limitation of containers and screen capture?

Unfortunately it is limited to get signals from the host system in the Docker container



> Could it work with elevated privileges or anything?

I do not think so

**Comment by Rickexe** [comment]
Hello,

I have a question about the releases. When I extract the file and run it on my coreelec device it doesn't say v2.0.0 alpha 4. It just says v2.0.0. Can you explain me why this is? I tried compiling my own tar.gz for amlogic but it still doesn't work. 

Thanks in advance.

**Comment by RickDB** [comment]
Agree with @Paulchen-Panther that a docker wouldn't really work due to the nature of screen capture in general as that needs direct tie-ins with display and access to LEDS  via SPI / USB serial.

It will work for some cases like AtmoOrb where you only need to receive and send TCP/UDP commands but we couldn't maintain a docker image just for those as it would create confusion for sure :)

**Comment by Hukuma1** [comment]
> It will work for some cases like AtmoOrb where you only need to receive and send TCP/UDP commands but we couldn't maintain a docker image just for those as it would create confusion for sure :)

Now it makes sense! Thank you.

I am using a great utility for Android TV (https://play.google.com/store/apps/details?id=com.abrenoch.hyperiongrabber) which captures the screen and sends data to Hyperion. In my use case Hyperion would work just fine in a Docker since it does not grab the screen itself, correct?

Anyway, keep up the great work in Hyperion.NG! Really takes media to another level.

**Comment by RickDB** [comment]
> 
> I am using a great utility for Android TV (https://play.google.com/store/apps/details?id=com.abrenoch.hyperiongrabber) which captures the screen and sends data to Hyperion. In my use case Hyperion would work just fine in a Docker since it does not grab the screen itself, correct?
> 

It would still need something that connects to the LEDs.
That can be via Serial interface (COM port) as Docker can forward that to the container, not sure about performance though as there will be an added layer.

**Comment by Hukuma1** [comment]
My setup is:
Android TV -> Arduino Nano -> RPi4 -> Hyperion.NG

Would running HNG in Docker work for this scenario? Guess I could always try and see but I haven't fully learned how to create a Docker container yet. D:

**Comment by RickDB** [comment]
If the Arduino Nano is the one driving the leds and its connected to the RPi4 via USB that should work, you need to lookup the /dev/ttyXX serial device and forward that to the docker container:

https://stackoverflow.com/questions/24225647/docker-a-way-to-give-access-to-a-host-usb-or-serial-device

Then again running in natively on the RPi4 would probably be easier though unless you are gonna use the RPi4 for different applications which docker is great for.


Side note:

Recently went full Linux for my home server and 95% of my server applications run in Docker containers, manage everything with Portainer (web based Docker "interface") and can highly recommend that one as it beats tinkering with CLI commands :)

Docker has a very bright future and makes for easy managing of just about everything you can think off so it's worth investing time into for sure.

**Comment by Hukuma1** [comment]
> Recently went full Linux for my home server and 95% of my server applications run in Docker containers, manage everything with Portainer (web based Docker "interface") and can highly recommend that one as it beats tinkering with CLI commands :)
> 
> Docker has a very bright future and makes for easy managing of just about everything you can think off so it's worth investing time into for sure.

Yep! Exactly got this right now. DietPi + Home Assistant + Portainer = unlimited possibilities.

So this is why I was hoping someone could create a Docker container for Hyperion.NG. The few I found were not for arm architecture.

**Comment by RickDB** [comment]
Too bad this one didn't publish the source code as that way we could try and run a test build on an ARM device:

https://hub.docker.com/r/sirfragalot/hyperion.ng

However you could mail him and ask to compile an image on a Raspberry or publish the Docker code to Github at least:

`LABEL MAINTAINER=jonwilliams84@gmail.com DESCRIPTION=Docker container for Hyperion-NG IMAGENAME=hyperion.ng OS=linux VERSION=0.1.0`

**Comment by Hukuma1** [comment]
https://hub.docker.com/layers/sirfragalot/hyperion.ng/0.1.0/images/sha256-c11bd96d1799ff631f1dcbffd2e9c5af7947b6e561cff7bffee87eacfcd26db6?context=explore

Does this show us the code? It's the makeup of the file.

**Comment by RickDB** [comment]
Partially yes, missing a few definitions but in theory can reverse engineer it off that.
Here's an example of valid DockerFile:

https://github.com/RickDB/Docker-PlexAniSync/blob/master/Dockerfile

**Comment by bradsjm** [comment]
This is a very simple Dockerfile for the PI if no local system is needed (e.g. using network video capture and WLED etc.)
```
FROM balenalib/raspberrypi3:run

RUN install_packages \
	libpython3.5 \
	libjpeg62-turbo \
	libavahi-compat-libdnssd1 \
	libdbus-1-3 \
	libqt5sql5 \
	libqt5serialport5 \
	libqt5network5 \
	libqt5widgets5 \
	libqt5gui5 \
	libqt5core5a

ENV RELEASE 2.0.0-alpha.5/Hyperion-2.0.0-alpha.5-Linux-armv7hf-rpi

RUN curl -fsSL https://github.com/hyperion-project/hyperion.ng/releases/download/${RELEASE}.deb > /tmp/hyperion.deb && dpkg -i /tmp/hyperion.deb && rm -f /tmp/hyperion.deb

EXPOSE 8090 8091 19333 19400 19445
VOLUME /config

CMD hyperiond -u /config
```

Build with: **docker build -t hyperion.ng .**

Run with: **docker run -d -v /path/on/host:/config --network host hyperion.ng**

**Comment by Hukuma1** [comment]
Thank you very much for this, @bradsjm!

It seems to be *almost* working.

`[hyperiond DISPMANXGRABBER] <ERROR> Snapshot failed: -1,
[hyperiond DISPMANXGRABBER] <ERROR> Previous line repeats 9 times,
[hyperiond LEDDEVICE] <DEBUG> <LedDeviceAdalight.cpp:83:receivedData()> >>received 4 bytes data,`

I am able to interface with Hyperion and change effect colors (since grabber not working it's in idle mode). But the Android Video Grabber is not able to connect it seems. Grabber options set to auto in settings for Hyperion. Any idea if some install file is missing perhaps?

**Comment by Hukuma1** [comment]
`FROM balenalib/raspberrypi3:run

RUN install_packages \
	libpython3.5 \
	libjpeg62-turbo \
	libavahi-compat-libdnssd1 \
	libdbus-1-3 \
	libqt5sql5 \
	libqt5serialport5 \
	libqt5network5 \
	libqt5widgets5 \
	libqt5gui5 \
	libqt5core5a

ENV RELEASE 2.0.0-alpha.7/Hyperion-2.0.0-alpha.7-Linux-armv7l
            
RUN curl -fsSL https://github.com/hyperion-project/hyperion.ng/releases/download/${RELEASE}.deb > /tmp/hyperion.deb && dpkg -i /tmp/hyperion.deb && rm -f /tmp/hyperion.deb
               
EXPOSE 8090 8091 19333 19400 19445
VOLUME /config

CMD hyperiond -u /config`

Trying to run new container with updated alpha 7 build and get an error "cannot read properly 'url' of undefined." Any ideas why? I only changed the env from alpha 5 to alpha 7 in the above code. Thanks in advance, @bradsjm!

**Comment by Hukuma1** [comment]
Nevermind. Looks to be a Portainer bug with Home Assistant. Probably works just fine once it gets fixed. Sorry for the commotion. :) Will leave this here in case anyone has similar issues. https://github.com/portainer/portainer/issues/4125

**Comment by Lord-Grey** [comment]
@Hukuma1 I assume you get the error from hyperion's UI,correct? I suggest you remove cookies and remove the browser cache.

**Comment by panec** [comment]
Had anyone had any problems with exposed ports? While the 8090 works fine, the 19333 19400 19445 cannot be accessed at all (tested from an android phone using Android grabber & Hyperion app)?

**Comment by panec** [comment]
I still have no clue why it is not working. This is a netstat from the running container with hyperion.ng
```
root@dba1ceb5-hyperion:/# netstat -antp
Active Internet connections (servers and established)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name    
tcp        0      0 127.0.0.11:40097        0.0.0.0:*               LISTEN      -                   
tcp        0      0 172.30.33.5:40174       151.101.0.204:80        TIME_WAIT   -                   
tcp        0      0 172.30.33.5:37554       151.101.62.132:80       TIME_WAIT   -                   
tcp        0      0 172.30.33.5:37524       151.101.62.132:80       TIME_WAIT   -                   
tcp6       0      0 :::19400                :::*                    LISTEN      196/hyperiond       
tcp6       0      0 :::19444                :::*                    LISTEN      196/hyperiond       
tcp6       0      0 :::19445                :::*                    LISTEN      196/hyperiond       
tcp6       0      0 :::8090                 :::*                    LISTEN      196/hyperiond       
tcp6       0      0 :::8092                 :::*                    LISTEN      196/hyperiond       
tcp6       0      0 172.30.33.5:8090        192.168.1.110:61592     ESTABLISHED 196/hyperiond       
tcp6       0      0 172.30.33.5:19444       192.168.1.75:51028      ESTABLISHED 196/hyperiond       
tcp6       0      0 172.30.33.5:8090        192.168.1.110:61587     ESTABLISHED 196/hyperiond       
tcp6       0      0 172.30.33.5:8090        192.168.1.110:61586     ESTABLISHED 196/hyperiond       
tcp6       0      0 172.30.33.5:8090        192.168.1.110:61585     ESTABLISHED 196/hyperiond  
```

`172.30.33.5` - is hassio IP address
`192.168.1.75` - is host machine IP address
` 192.168.1.110` - is the address of the PC from which I access the web GUI

Any suggestions why I cannot connect to 19444 or 19445?

**Comment by BurninTurtles** [comment]
I just want to throw this out there. I was able to get some dockerfile (I wish I could find it again) for this running and by docker-compose by using the below elements in the compose file. So a docker setup is totally possible.

`Devices:
    /dev/video0:/dev/video0
    /dev/spidev0.0:/dev/spidev0.0
    /dev/spidev0.1:/dev/spidev0.1`

my only issues were related to my capture device at `video0` which I'm 95% sure are an artifact of my setup and cheap hardware. I recently rebuilt my pi and neglected to make a backup, so I will update this once I find the dockerfile I used.
